### PR TITLE
Validations for unused intents and utterances

### LINKED ISSUE #10079: Data validation with --fail-on-warnings fails but there are no warnings printed
### Rasa Open Source version

2.8.6

### Rasa SDK version

2.8.2

### Rasa X version

None

### Python version

3.8

### What operating system are you using?

OSX

### What happened?

Output for `rasa data validate`:

```
2021-11-03 16:22:58 INFO     rasa.validator  - Validating intents...
2021-11-03 16:22:58 INFO     rasa.validator  - Validating uniqueness of intents and stories...
2021-11-03 16:22:58 INFO     rasa.validator  - Validating utterances...
2021-11-03 16:22:58 INFO     rasa.validator  - Story structure validation...
Processed story blocks: 100%|█████████████████████████████████████████████████| 38/38 [00:00<00:00, 365.67it/s, # trackers=1]
2021-11-03 16:22:58 INFO     rasa.core.training.story_conflict  - Considering all preceding turns for conflict analysis.
2021-11-03 16:23:12 INFO     rasa.validator  - No story structure conflicts found.
```

Output for `rasa data validate --fail-on-warnings`:

```
2021-11-03 15:35:55 INFO     rasa.validator  - Validating intents...                                                                                                                                                                                        
2021-11-03 15:35:55 INFO     rasa.validator  - Validating uniqueness of intents and stories...                                                                                                                                                              
2021-11-03 15:35:56 INFO     rasa.validator  - Validating utterances...                                                                                                                                                                                     
Project validation completed with errors.
```

There are no warnings printed for both commands but the one with `--fail-on-warnings` fails.

When running with `-vv` option I can see `DEBUG` level logs (but still no warnings) like:
```
2021-11-03 17:36:21 DEBUG    rasa.validator  - The intent 'back' is listed in the domain file, but is not found in the NLU training data.
```
and
```
2021-11-03 17:36:21 DEBUG    rasa.validator  - The utterance 'utter_default' is not used in any story.
```

Possibly related source code:
* https://github.com/RasaHQ/rasa/blob/2.8.6/rasa/validator.py#L137
* https://github.com/RasaHQ/rasa/blob/2.8.6/rasa/validator.py#L196

### Command / Request

```shell
rasa data validate --fail-on-warnings
```

### Relevant log output

```shell
2021-11-03 15:35:55 INFO     rasa.validator  - Validating intents...                                                                                                                                                                                        
2021-11-03 15:35:55 INFO     rasa.validator  - Validating uniqueness of intents and stories...                                                                                                                                                              
2021-11-03 15:35:56 INFO     rasa.validator  - Validating utterances...                                                                                                                                                                                     
Project validation completed with errors.
```

**Definition of done**
- [ ] output for `rasa data validate --fail-on-warnings` correct

### PR DESCRIPTION
**Pull request by losterloh**
Fixes #10079.

The following changes were made:
- we exclude the default / implicit intents from generating warnings if they're unused; otherwise, any empty / newly initialized rasa project would throw warnings for unused intents, which would be confusing to the user
- if an intent is defined, but not used in the NLU data at all, we generate a warning and consequently abort if `--fail-on-warnings`.
- if an intent is defined, but not used either in a story or rule we generate a warning and consequently abort if `--fail-on-warnings`. Note here that the debug message that was printed before was also inexact since it mentioned only stories, this was clarified to also include rules.
- The same as in the bullet point before was also implemented for utterances

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/validator.py [hunk 1]
```diff
@@ -64,9 +65,14 @@ def verify_intents(self, ignore_warnings: bool = True) -> bool:
 
         nlu_data_intents = {e.data["intent"] for e in self.intents.intent_examples}
 
-        for intent in self.domain.intents:
+        # Exclude the default intents that are always implicitly present
+        for intent in [
+            item
+            for item in self.domain.intents
+            if item not in constants.DEFAULT_INTENTS
+        ]:
```
Comment (wochinge) [inline rasa/validator.py:73]: optional: We could also make this a property on `Domain` to avoid duplicating the logic

File: data/test_validation/data/stories.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+version: "2.0"
+stories:
+- story: happy path
+  steps:
+  - intent: greet
+  - action: utter_greet
+
```
Comment (wochinge) [inline data/test_validation/data/stories.yml:7]:
```suggestion
```

File: tests/test_validator.py [hunk 3]
```diff
@@ -213,6 +226,44 @@ def test_verify_there_is_example_repetition_in_intents(nlu_data_path: Text):
     assert not validator.verify_example_repetition_in_intents(False)
 
 
+def test_verify_logging_message_for_intent_not_used_in_nlu(
+    caplog, validator_under_test
```
Comment (wochinge) [inline tests/test_validator.py:230]: can we please add type annotations (same for the others)?

File: rasa/validator.py [hunk 4]
```diff
@@ -64,9 +65,14 @@ def verify_intents(self, ignore_warnings: bool = True) -> bool:
...
             if intent not in nlu_data_intents:
-                logger.debug(
+                rasa.shared.utils.io.raise_warning(
```
Comment (wochinge) [inline rasa/validator.py:75]: I've noticed that there are a couple of others with `debug` messages. In my opinion we should be consistent and have them all as warnings, what do you think?

**Review by wochinge — COMMENTED** [review]
Welcome to Rasa Open Source @losterloh 🚀

File: tests/test_validator.py (hunk 3 above)
Comment (losterloh) [reply tests/test_validator.py:230]: Yes; however, if that's what we want, and everywhere, I suggest we should enforce this with a linter.

File: rasa/validator.py (hunk 4 above)
Comment (losterloh) [reply rasa/validator.py:75]: Yes, I've noticed the same thing, but I feel like not all of them are clear cut - which actually  should be more of a warning, and which are really purely for debug purposes? Happy to go through with somebody who's in deeper already than me.

File: rasa/validator.py (hunk 1 above)
Comment (losterloh) [reply rasa/validator.py:73]: Yup can definitely do! Do we have a general policy for code repetitions and when to de-duplicate them? My usual approach is to tackle those when code would otherwise be triplicated.

File: data/test_validation/data/stories.yml (hunk 2 above)
Comment (losterloh) [reply data/test_validation/data/stories.yml:7]: Whoops! You're right of course.

File: tests/test_validator.py (hunk 3 above)
Comment (losterloh) [reply tests/test_validator.py:230]: For instance, in this file, the `caplog` that is used to capture log output is not annotated anywhere, the type would even need to be imported into the file first from a private package. I'll do this now.

File: rasa/validator.py (hunk 4 above)
Comment (wochinge) [reply rasa/validator.py:77]: Thanks for changing the to warning as well 👍🏻 Makes way more sense now 🎉

File: tests/test_validator.py [hunk 5]
```diff
@@ -213,6 +227,46 @@ def test_verify_there_is_example_repetition_in_intents(nlu_data_path: Text):
...
+
+def test_verify_logging_message_for_unused_utterance(
+    caplog: LogCaptureFixture, validator_under_test: Text
+):
+    caplog.clear()
+    with pytest.warns(UserWarning) as record:
+        validator_under_test.verify_utterances_in_stories(False)
+
+    for m in record:
+        print(m)
```
Comment (wochinge) [inline tests/test_validator.py:264]:
```suggestion
```

File: data/test_config/config_defaults.yml [hunk 6]
```diff
@@ -24,3 +24,16 @@ pipeline: []
...
+# # See https://rasa.com/docs/rasa/policies for more information.
+#   - name: MemoizationPolicy
+#   - name: RulePolicy
+#   - name: UnexpecTEDIntentPolicy
+#     max_history: 5
+#     epochs: 100
+#   - name: TEDPolicy
+#     max_history: 5
+#     epochs: 100
+#     constrain_similarities: true
```
Comment (wochinge) [inline data/test_config/config_defaults.yml:39]: is this change actually needed?

**Review by wochinge — COMMENTED** [review]
> Yes; however, if that's what we want, and everywhere, I suggest we should enforce this with a linter.

Good point 👍🏻 We do this for the production code but are less strict in the test code 😬

File: tests/test_validator.py (hunk 5 above)
Comment (losterloh) [reply tests/test_validator.py:264]: 🤦  thanks. Another thing where we could think about involving the linter 😓

Comment (wochinge) [reply tests/test_validator.py]: There is actually a https://pypi.org/project/flake8-print/ 👍🏻 Will add create an issue for it 👍🏻

Comment (losterloh) [reply tests/test_validator.py]: Perfect! 🎉

Comment (wochinge) [reply tests/test_validator.py]: https://github.com/RasaHQ/rasa/issues/10205

File: rasa/validator.py (hunk 1 above)
Comment (wochinge) [reply rasa/validator.py]: > My usual approach is to tackle those when code would otherwise be triplicated.

I think that's a good rule of thumb. In this case I just felt like we'd need that in several places sooner or later and it felt quite "natural" to the Domain. I like the solution which you chose 💯

File: data/test_config/config_defaults.yml (hunk 6 above)
Comment (losterloh) [reply data/test_config/config_defaults.yml:39]: I think it happens around here:
https://github.com/RasaHQ/rasa/blob/d13b153574acb1a70dd4c170a8bc23b514e48a63/rasa/shared/importers/autoconfig.py#L67

Comment (wochinge) [reply data/test_config/config_defaults.yml:39]: Mhm, I see your point. We should probably change the tests to not write there instead 🙈 Fine to keep it as it is 👍🏻

Comment (wochinge) [reply data/test_config/config_defaults.yml:39]: yep, but in my opinion the tests should e.g. copy the config before doing this.

File: tests/test_validator.py [hunk 7]
```diff
@@ -213,6 +227,44 @@ def test_verify_there_is_example_repetition_in_intents(nlu_data_path: Text):
...
+    with pytest.warns(UserWarning) as record:
+        validator_under_test.verify_intents_in_stories(False)
+
+    assert "The intent 'goodbye' is not used in any story or rule." in (
+        m.message.args[0] for m in record
+    )
+
+
+def test_verify_logging_message_for_unused_utterance(
+    caplog: LogCaptureFixture, validator_under_test: Text
```
Comment (wochinge) [inline tests/test_validator.py:257]:
```suggestion
    caplog: LogCaptureFixture, validator_under_test: Validator
```

File: tests/test_validator.py [hunk 8]
```diff
@@ -213,6 +227,44 @@ def test_verify_there_is_example_repetition_in_intents(nlu_data_path: Text):
...
+
+    assert (
+        "The intent 'goodbye' is listed in the domain file, "
+        "but is not found in the NLU training data."
+        in (m.message.args[0] for m in record)
+    )
+
+
+def test_verify_logging_message_for_intent_not_used_in_story(
+    caplog: LogCaptureFixture, validator_under_test: Text
```
Comment (wochinge) [inline tests/test_validator.py:245]:
```suggestion
    caplog: LogCaptureFixture, validator_under_test: Validator
```

File: tests/test_validator.py [hunk 9]
```diff
@@ -213,6 +227,44 @@ def test_verify_there_is_example_repetition_in_intents(nlu_data_path: Text):
     assert not validator.verify_example_repetition_in_intents(False)
 
 
+def test_verify_logging_message_for_intent_not_used_in_nlu(
+    caplog: LogCaptureFixture, validator_under_test: Text
```
Comment (wochinge) [inline tests/test_validator.py:231]:
```suggestion
    caplog: LogCaptureFixture, validator_under_test: Validator
```

**Review by wochinge — APPROVED** [review]
💯 Congrats to your first contribution to Rasa Open Source! For context: I'd usually target 2.8.x for this change, but as we are very close to the 3.0 release and this issue is not very urgent it's in my opinion perfectly fine to merge to `main`.

File: data/test_config/config_defaults.yml (hunk 6 above)
Comment (losterloh) [reply data/test_config/config_defaults.yml:39]: Yes, that would be another way to do it; in general, I wonder if just writing over an existing file is really what we want. I'm assuming that in most cases - although not here - this would happen in a more interactive context anyways, where it might also make sense, for instance, to ask the user if they want these default comments to be added.

Comment (wochinge) [reply data/test_config/config_defaults.yml:39]: In my opinion fine in _this_ case as we're just adding a comment and not actually changing something 🤔 Agree with you in general though 👍🏻

File: tests/test_validator.py (hunk 9 above)
Comment (losterloh) [reply tests/test_validator.py]: Thanks! Fixed.

**Comment by losterloh** [comment]
thanks @wochinge! I'll merge this then once CI runs through again after the latest small change (& rebase against `main`).

**Comment by wochinge** [comment]
Congrats on beating the CI! 😁
